ruby/ql/lib/codeql/ruby/frameworks/data/internal/ApiGraphModelsSpecific.qll

```diff
@@ -0,0 +1,165 @@
+ * API::Node getExtraSuccessorFromNode(API::Node node, AccessPathToken token)
+ * API::Node getExtraSuccessorFromInvoke(InvokeNode node, AccessPathToken token)
+ * predicate invocationMatchesExtraCallSiteFilter(InvokeNode invoke, AccessPathToken token)
+ * InvokeNode getAnInvocationOf(API::Node node)
+ * ```
+ */
+
+private import ruby
+private import codeql.ruby.DataFlow
+private import codeql.ruby.dataflow.internal.DataFlowPrivate as DataFlowPrivate
+private import ApiGraphModels
+
+class Unit = DataFlowPrivate::Unit;
+
+// Re-export libraries needed by ApiGraphModels.qll
+import codeql.ruby.ApiGraphs
+import codeql.ruby.dataflow.internal.AccessPathSyntax as AccessPathSyntax
+private import AccessPathSyntax
+
+/**
+ * Holds if models describing `package` may be relevant for the analysis of this database.
+ *
+ * In the context of Ruby, this is the name of a Ruby gem.
+ */
+bindingset[package]
+predicate isPackageUsed(string package) {
+  // For now everything is modelled as an access path starting at any top-level, so the package name has no effect.
+  //
+  // We allow an arbitrary package name so that the model can record the name of the package in case it's needed in the future.
+  //
+  // In principle we should consider a package to be "used" if there is a transitive dependency on it, but we can only
+  // reliably see the direct dependencies.
+  //
+  // In practice, packages try to use unique top-level module names, which mitigates the precision loss of not checking
+  // the package name.
+  any()
+}
+
+/** Gets a Ruby-specific interpretation of the `(package, type, path)` tuple after resolving the first `n` access path tokens. */
+bindingset[package, type, path]
+API::Node getExtraNodeFromPath(string package, string type, AccessPath path, int n) {
+  isRelevantFullPath(package, type, path) and
+  package = any(string s) and // Allow any package name, see `isPackageUsed`.
+  type = "" and
+  n = 0 and
+  result = API::root()
+  or
+  // A row of form `;any;Method[foo]` should match any method named `foo`.
+  package = any(string s) and
+  type = "any" and
+  n = 1 and
+  exists(EntryPointFromAnyType entry |
+    methodMatchedByName(path, entry.getName()) and
+    result = entry.getNode()
+  )
+}
+
+/**
+ * Holds if `path` occurs in a CSV row with type `any`, meaning it can start
+ * matching anywhere, and the path begins with `Method[methodName]`.
+ */
+private predicate methodMatchedByName(AccessPath path, string methodName) {
+  isRelevantFullPath(_, "any", path) and
+  exists(AccessPathToken token |
+    token = path.getToken(0) and
+    token.getName() = "Method" and
+    methodName = token.getAnArgument()
+  )
+}
+
+/**
+ * An API graph entry point corresponding to a method name such as `foo` in `;any;Method[foo]`.
+ *
+ * This ensures that the API graph rooted in that method call is materialized.
+ */
+private class EntryPointFromAnyType extends API::EntryPoint {
```

Comment:
Shouldn't we make sure to import this file, or `ApiGraphModels.qll` into `ApiGraph.qll` to avoid recomputation?

Reply:
Sure, if you want.

In JS we try to ensure all files that aren't query-specific are transitively included from `javascript.qll` (but not necessarily made visible). Is there a reason you're not doing this in Ruby? Seems more robust IMO.

Reply:
As long as they are not made visible that should work too. However, I don't know if that makes compilation slower.